Added code and test to enable vxlan MAC if present in node spec.

tathagatachowdhury opened the pull request:
## Description
<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
-->

## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests (delete as appropriate) In plan/Not needed/Done
- [ ] Documentation
- [ ] Backport
- [ ] Release note

## Release Note
<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
None required
```

marvin-tigera committed:
Automated Typha pin update

marvin-tigera committed:
Automated Typha pin update

fasaxc commented on config/config_params.go:
```diff
@@ -170,6 +170,7 @@ type Config struct {
 	VXLANVNI            int    `config:"int;4096"`
 	VXLANMTU            int    `config:"int;1410;non-zero"`
 	IPv4VXLANTunnelAddr net.IP `config:"ipv4;"`
+	MacVXLANTunnelAddr  string `config:"string;"`
```
This should be `VXLANTunnelMACAddr`; hard to explain why, `MacVXLANTunnelAddr` just sounds backwards!

tathagatachowdhury replied:
Acknowledged and changed.

fasaxc commented on calc/vxlan_resolver.go:
```diff
@@ -483,6 +537,14 @@ func (c *VXLANResolver) determineGatewayForRoute(r vxlanRoute) string {
 func (c *VXLANResolver) vtepMACForHost(nodename string) string {
+	mac, ok := c.nodeNameToVXLANMac[nodename]
+
+	if ok {
```
Don't need the extra "ok" check here.  If value isn't present, map will return `""`

tathagatachowdhury replied:
Acknowledged and changed.

fasaxc commented on calc/vxlan_resolver.go:
```diff
@@ -263,6 +265,58 @@ func (c *VXLANResolver) OnHostConfigUpdate(update api.Update) (_ bool) {
+				}
+
+				// We've already sent a VTEP for this node, and the node's MAC address has changed.
+				// We need to revoke it and any routes before sending any updates.
```
As discussed, for the MAC address case, there's no need to revoke the routes since the MAC only affects the VTEP.

fasaxc replied:
I think that means this whole block can be simplified a lot vs the IP case.  The MAC isn't required to send the VTEP in the same way that the IP address is required so adding/removing the MAC can never invalidate the routes, it can only update the VTEP (if and only if the VTEP has already been sent).

tathagatachowdhury replied:
I have updated the code.

fasaxc commented on calc/vxlan_resolver.go:
```diff
@@ -263,6 +265,58 @@ func (c *VXLANResolver) OnHostConfigUpdate(update api.Update) (_ bool) {
+				c.kickPendingRoutes(pendingSet)
+			}
+		} else {
+			// Withdraw any routes which target this VTEP, followed by the VTEP itself.
```
This is not correct.  When the MAC is removed, we should revert to the auto-generated MAC address

tathagatachowdhury replied:
Changed the code.

fasaxc commented on calc/states_for_test.go:
```diff
@@ -906,6 +906,33 @@ var vxlanWithBlock = empty.withKVUpdates(
 )
 
+// Minimal VXLAN set-up with a tunnel IP and MAC address.
+var vxlanWithBlockAndMAC = empty.withKVUpdates(
```
You could define this more simply as 
```
var vxlanWithBlockAndMAC = vxlanWithBlock.withKVUpdates(
	KVPair{Key: remoteHostVXLANTunnelMACConfigKey, Value: remoteHostVXLANTunnelMAC},
).withName("VXLAN MAC").withVTEPs(
	// VTEP for the remote node.
		//Mac:            remoteHostVXLANTunnelMAC,
		//Mac:            "66:3e:ca:a4:db:65",
	proto.VXLANTunnelEndpointUpdate{
		Node:           remoteHostname,
		Mac:            remoteHostVXLANTunnelMAC,
		Ipv4Addr:       remoteHostVXLANTunnelIP,
		ParentDeviceIp: remoteHostIP.String(),
	},
)
```

tathagatachowdhury replied:
Acknowledged and changed.

fasaxc commented on calc/states_for_test.go:
```diff
@@ -1027,6 +1054,10 @@ var vxlanTunnelIPDelete = vxlanWithBlock.withKVUpdates(
 
+var vxlanMACDelete = vxlanWithBlockAndMAC.withKVUpdates(
+	KVPair{Key: remoteHostVXLANTunnelMACConfigKey, Value: nil},
+).withName("VXLAN host tunnel MAC removed").withRoutes().withVTEPs()
```
After deleting the MAC, it should go back to using the auto-generated MAC so there should be routes and VTEPs.

The calc graph's invariant is that if you send updates A and B and then send a delete for B then the resulting state should go back to A.

tathagatachowdhury replied:
Acknowledged and changed.

fasaxc committed:
Improve calc graph test infra:
- avoid adding empty state to end of tests since it makes single-state
  async tests meaningless
- improve injector diags
- add env vars to quickly disable test expansion and async tests.

fasaxc committed:
Fix that routes were not re-sent if node IPs changed.

fasaxc committed:
Merge pull request #2030 from fasaxc/fix-vxlan

Fix that routes were not re-sent if node IPs changed.

Commit:
Added code and test to enable vxlan MAC if present in node spec.
